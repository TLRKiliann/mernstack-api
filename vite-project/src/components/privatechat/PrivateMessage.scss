.private--chat {
  position: absolute;
  top: 0px;
  right: 0px;
  bottom: 0px;
  left: 0px;
  background: linear-gradient(180deg, purple,
    orange);
  color: whitesmoke;
  z-index: 1;

  .title--privatechat {
    margin-top: 80px;
    margin-left: 20px;
    font-family: "Junction";
    text-shadow: 2px 1px 1px #000;
  }
}

.private--terminal {
  width: 70%;
  height: 70%;
  margin: auto;
  margin-top: 3%;
  display: flex;
  flex-direction: column;
  background-color: rgba(10,10,100, 0.6);
  border: 1px solid slategray;
  border-radius: 15px 15px 15px 15px;
}

.sub--privateterminal {
  width: auto;
  height: 70px;
  display: flex;
  align-items: center;
}

.flex--imgnameroom {
  width: 100%;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;

  background: #242424;
  border-bottom: 1px solid slategray;
  border-radius: 15px 15px 0px 0px;
}

.img--private {
  border-radius: 15px 0px 0px 0px;
}

.title--private {
  margin-left: 10px;
  font-size: 1.0rem;
  font-family: "Junction";
  background: linear-gradient(to right, #f32170,
  #eedd44);
  -webkit-text-fill-color: transparent;
  -webkit-background-clip: text;
}

.private--messagebox {
  border: none;
  width: auto;
  height: calc(100% - 90px);
  padding-left: 10px;
  overflow: scroll;
  overflow-x: hidden;
}

.inputbtn--privateterminal {
  border-top: 1px solid slategray;
  display: flex;
  align-items: center;
  justify-content: space-evenly;
  padding: 10px;
  border-radius: 0px 0px 15px 15px;
  input {
    width: 85%;
    padding: 3px 10px;
  }
  button {
    width: 10%;
    padding: 3px 10px;
    margin-left: 10px;
  }
}