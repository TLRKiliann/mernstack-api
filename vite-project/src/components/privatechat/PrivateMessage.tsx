import React, {useState, useEffect} from 'react'
import { useParams } from 'react-router-dom'
import { useAuthLogin } from '../../context/AuthProvider'
import servicePrivate from '../../services/servicePrivate'
import { UserType } from '../../models/usertype'
import './PrivateMessage.scss'

const PrivateMessage: React.FC = () => {

  const params = useParams<{ retrieveRoom?: string }>()
  const [roomStyle, setRoomStyle] = useState<{params?: string}>(params.retrieveRoom)

  const dateOfTheDay: Date = Date()
  const {otherUser, username, versusUser} = useAuthLogin()

  const [displayUser, setDisplayUser] = useState<object>({})
  const [privateMsg, setPrivateMsg] = useState<string>("")
  const [privateSeveralMsg, setPrivateSeveralMsg] = useState<Array<string>>([])

  useEffect(() => {
    if (Object.keys(otherUser).length === 0) {
      console.log("no otherUser")
      setDisplayUser(versusUser)
    } else {
      console.log("no versusUser")
      setDisplayUser(otherUser)
    }
    //const intervalId = setInterval(() => {
    servicePrivate
      .getMsgPrivate()
      .then((response) => {
        setPrivateSeveralMsg(response)
      })
    return () => console.log("Updated ! (2)")
    //}, 1000)
    //return () => clearInterval(intervalId)
  }, [privateSeveralMsg])

  const generateId = () => {
    const maxId: number = privateSeveralMsg.length > 0
      ? Math.max(...privateSeveralMsg.map(msg => msg.id))
      : 0
    return maxId + 1
  }

  const handleUserMessage = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault()
    const privateRoom = "Private"
    if (privateMsg) {
      const msgPrivate = {
        id: generateId(),
        date: dateOfTheDay.slice(0, 24),
        user: `${username}`,
        msg: `${privateMsg} ✉`,
        room: `${privateRoom}`
      }

      servicePrivate
        .postMsgPrivate(msgPrivate)
        .then(initialData => {
          setPrivateSeveralMsg(privateSeveralMsg.concat(msgPrivate))
        })
        .catch((error) => {
          privateSeveralMsg([])
          console.log("Send msg error (Private-Chat)!")
        })
    } else {
      console.log("message === undefined")
    }
    setPrivateMsg("")
  }

  return(
    <div className="private--chat">

      <h1 className="title--privatechat">
        {roomStyle} Chat
      </h1>

      <div className="private--terminal">
        <div className="sub--privateterminal">
            <div className="flex--imgnameroom">
              <img
                src={displayUser.img}
                width="95px"
                height="65px"
                className="img--private"
                alt="no img"
              />
              <h4 className="title--private">{displayUser.firstName}</h4>
              <h4 className="title--private">{displayUser.lastName}</h4>
              <h4 className="title--private">{displayUser.age} years</h4>
              <h4 className="title--private">{displayUser.email}</h4>
              <h4 className="title--private">{displayUser.isConnected ? (
                <span
                  className="span--useronline"
                  style={{color: 'lightgreen'}}
                >
                  ✔
                </span>
                ) : (
                <span 
                  className="span--useronline">
                  ❌
                </span>
                )
              }
              </h4>
            </div>

        </div>

        <div className="private--messagebox">
          <div>
          {privateSeveralMsg?.map((data) => (
            <div key={data.id} className="map--msg">
              <p className="para--chat">$ ▶ {data.user} ~ {data.msg}</p>
              <legend className="legend--date">{data.date}</legend>
            </div>
            ))
          }
          </div>
        </div>

        <div className="inputbtn--privateterminal">
          <input
            type="text"
            value={privateMsg}
            onChange={(e) => setPrivateMsg(e.target.value)}
            placeholder="Enter your message here !"
          />
          <button
            type="button"
            onClick={handleUserMessage}
            className="private--btn"
          >
            Enter
          </button>
        </div>
      </div>

    </div>
  )
}

export default PrivateMessage
