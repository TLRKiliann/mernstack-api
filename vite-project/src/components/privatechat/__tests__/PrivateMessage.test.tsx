import React from "react"
import { MemoryRouter } from 'react-router-dom'
import { test, expect, vi } from 'vitest'
import {fireEvent, render, screen} from '@testing-library/react'
import { create } from 'react-test-renderer'
import '@testing-library/jest-dom'
import PrivateMessage from "../PrivateMessage.tsx"


test("PrivateMessage MatchSnapShot", () => {
  const treePrivate = create(
    <MemoryRouter>
      <PrivateMessage />
    </MemoryRouter>
  )
  expect(treePrivate.toJSON()).toMatchSnapshot()
})

test('PrivateMessage testid', () => {
  render(<PrivateMessage />)
  const mytestid = screen.getByTestId("terminaltest")
  expect(mytestid).toBeInTheDocument()
})

test('PrivateMessage test by text', () => {
  render(<PrivateMessage />)
  const mytestid = screen.getByText(/chat/i)
  expect(mytestid).toBeInTheDocument()
})

test("PrivateMessage click btn get test", () => {
  const handleUserMessage = vi.fn()
  const { getByTestId } = render(
    <button onClick={handleUserMessage}
      data-testid="private--btn">
    </button>)
  fireEvent.click(screen.getByTestId("private--btn"))
  expect(handleUserMessage).toHaveBeenCalledTimes(1)
})

//vi.setTimeout(7000)
test("Captures value from changeMessage() PrivateMessage", async() => {
  function changeMessage(evt) {
  expect(evt.target.value).toEqual("valuechange")
  }
  const { getByTestId } = render(
    <PrivateMessage
      onChange={changeMessage} 
      data-testid="inputTest"
    />
  )
  const node = getByTestId("inputTest")
  fireEvent.change(node, { target: { value: "valuechange" } })
})