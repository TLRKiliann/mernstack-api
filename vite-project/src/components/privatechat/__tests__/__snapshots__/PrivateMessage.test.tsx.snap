// Vitest Snapshot v1

<<<<<<< HEAD
exports[`MatchSnapShot test PrivateMessage 1`] = `
<div
  className="private--chat"
=======
exports[`PrivateMessage MatchSnapShot 1`] = `
<div
  className="private--chat"
  data-testid="terminaltest"
>>>>>>> master
>
  <h1
    className="title--privatechat"
  >
     Chat
  </h1>
  <div
    className="private--terminal"
  >
    <div
      className="sub--privateterminal"
    >
      <div
        className="flex--imgnameroom"
      >
        <img
          alt="no img"
          className="img--private"
          height="64px"
          width="95px"
        />
      </div>
      <div
        className="title--private"
      >
        <h4
          className="subtitle--private"
        />
        <h4
          className="subtitle--private"
        />
        <h4
          className="subtitle--private"
        >
           years
        </h4>
        <h4
          className="subtitle--private email"
        />
        <h4
          className="lastsubtitle--private"
        >
          <span
            className="span--useronline"
          >
            ❌
          </span>
        </h4>
      </div>
    </div>
    <div
      className="private--messagebox"
    >
      <div />
    </div>
    <div
      className="inputbtn--privateterminal"
    >
      <input
<<<<<<< HEAD
=======
        data-testid="inputTest"
>>>>>>> master
        onChange={[Function]}
        placeholder="Enter your message here !"
        type="text"
        value=""
      />
      <button
        className="private--btn"
<<<<<<< HEAD
        data-testid="btn-privatemsg"
=======
>>>>>>> master
        onClick={[Function]}
        type="button"
      >
        Enter
      </button>
    </div>
  </div>
</div>
`;
